### ISSUE TITLE
Sample Image test is causing exception

### ISSUE THREAD

**Issue by babazaroni** [alicevision/CCTag#74]
I've built CCTag on a Ubuntu 16.04 machine using OpenCV 3.3.1, cpu only.  However, the sample image test fails in Identification.cpp:1206 at `assert ( outerPoints.size() >= 5);` line. 

The problem appears to be that `numerical::ellipseFitting` in Multiresolution.cpp is causing an exception every call so that outerPoints.size() is always zero.

Any ideas on how to resolve this problem?  Should I be using a previous CCTag version?

**Comment by simogasp** [comment]
can you try the `specialCases` branch, it could be related to #66

**Comment by babazaroni** [comment]
Thanks for the quick reply.  Compiling the specialCases branch now.  I need to satisfy the `boot_unit_test_framework` requirement.

**Comment by babazaroni** [comment]
The specialCases branch works for me.  Thanks very much!

**Comment by simogasp** [comment]
great!

**Comment by GuillaumeHauss** [comment]
Hi guys, 
I've just tried the specialCases in CPU-only mode, and it still crashes.
Is it related to the CPU-only mode? or is it still an ongoing development? 

Thank you for your work, by the way !

**Comment by simogasp** [comment]
Hi @GuillaumeHauss 
thanks for your appreciation!
Does it crash at the same point? If it crashes during the ellipse fitting like above it's not a problem of the CPU-only mode, as that step is always done on CPU.
Just to complete the information, you are running the `detection` application, is that so? Are you using an image, and if so can you share it so we can try to replicate the problem?

**Comment by babazaroni** [comment]
The specialCases version that works for me is 273aa5a.  You might want to try that version.

**Comment by babazaroni** [comment]
The specialCases version that works for me is d22fd2ea.  You might want to try that version.

**Comment by GuillaumeHauss** [comment]
Hi guys, 
Thank you for this quick answer! Here are the answers to your inquiries
- It does crash at the same point, when doing the `assert(outerPoints.size()>=5)`
- I'm running the detection application, with the command `build/src/detection -n 3 -i sample/01.png` .
- I've tried on both samples images, the crash is the same

I'm gonna try your version,; @babazaroni , thank you!

**Comment by simogasp** [comment]
@GuillaumeHauss @babazaroni 

So looking into the code it turned out I was wrong, that is not related to the `specialCases` branch.
I guess you are building in **Debug** mode, you should try **Release** mode (`-DCMAKE_BUILD_TYPE:STRING=Release`) which will (of course) skip that assert and complete correctly the computation, as it happens to me. (BTW Release mode is incidentally way much faster)

That does not solve the issue at hand, obviously. I'm not the expert of the algorithm so we need @griwodz and @lcalvet to jump in...
I see that line has been switched back and forth between an `assert()` and  a simple `if` that exits the function with a code error if the condition is not satisfied.
At the beginning there was an `assert`
https://github.com/alicevision/CCTag/commit/a10fce2d3dc1d9985955df6b49996063909548a7
that has been changed for a if-return condition and then again rolled back
https://github.com/alicevision/CCTag/commit/849a62f0ebf19d8af979c0bd7f78fcd494e33e46
as it is today.

So the question is, why preferring an assert at that point? Should the previous function call `getSortedOuterPoints()` ALWAYS return more than 5 points? My guess it is not, as it is some sort of filtering so it could happen, so the `if` would be more appropriate IMHO.
But maybe (or surely... :-) ) there is something that I'm missing.

**Comment by GuillaumeHauss** [comment]
@simogasp 

it worked like a charm! It's only the CPU-version though, I need to figure out why it's not compiling against the right CUDA library. 
But I got an output, and markers have been detected. 
Here are a few remarks:
- for sample image 01.png, I got 4 markers identified, and a computing time of 0.74s
- for sample image 02.png, I got 6 markers identified and a computing time of 0.68s
Does that seems familiar to you? because it's quite slow for a real-time application :) 

Thank you very much for your help by the way !

**Comment by griwodz** [comment]
Hi @simogasp and @GuillaumeHauss,

Sorry, my everyday machine is still Ubuntu 16.04 and CUDA 8. I gave it a brief try with Ubuntu 16.04 and CUDA 9 not so long ago and I ran into a problem with the combination of a recent Boost version and CUDA 9. Unfortunately, I didn't have the time to follow up yet.

**Comment by GuillaumeHauss** [comment]
Hi all, 
I successfully installed CUDA 8.0 besides CUDA 9.0, and change the symlink /usr/local/cuda to point at the right version, so that CCTag will compile again CUDA 8.0

So far so good! However, when compiling, I got the following crash
```
In function ‘void cctag::cctagDetection(cctag::CCTag::List&, int, std::size_t, const cv::Mat&, const cctag::Parameters&, const cctag::CCTagMarkersBank&, bool, cctag::logtime::Mgmt*)’:
/home/guillaume/Documents/99_Utils/CCTag/src/cctag/Detection.cpp:829:16: error: ‘class cctag::TagPipe’ has no member named ‘debug’
         pipe1->debug( pix, params );
```
I looked it up in the file src/cuda/tag.h and indeed there's no function prototype `debug` in the class `cctag::TagPipe`

I guess that's another issue; so i'll just comment out this line for now :)

**Comment by simogasp** [comment]
thanks for reporting that! It looks like a leftover of a previous refactoring and Travis only tests the release mode...

**Comment by simogasp** [comment]
closing as it has been fixed by #86 (assert) and #92 (obsolete debug function)
